include/algorithms/k_nearest_neighbors/bf_knn_classification_training_batch.h

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Constructs a BF kNN prediction algorithm by copying input objects and parameters
+     * of another BF kNN prediction algorithm
+     * \param[in] other Algorithm to use as the source to initialize the input objects
+     *                  and parameters of the algorithm
+     */
+    Batch(const Batch<algorithmFPType, method> & other);
+
+    /**
+    * Gets parameter of the algorithm
+    * \return parameter of the algorithm
+    */
+    virtual ParameterType& parameter() { return *static_cast<ParameterType*>(_par); }
+
+    /**
+    * Gets parameter of the algorithm
+    * \return parameter of the algorithm
+    */
+    virtual const ParameterType& parameter() const { return *static_cast<const ParameterType*>(_par); }
+
+    /**
+     * Get input objects for the BF kNN prediction algorithm
+     * \return %Input objects for the BF kNN prediction algorithm
+     */
+    InputType * getInput() DAAL_C11_OVERRIDE { return static_cast<InputType*>(_in); }
+
+    /**
+     * Returns the method of the algorithm
+     * \return Method of the algorithm
+     */
+    virtual int getMethod() const DAAL_C11_OVERRIDE { return(int)method; }
+
+    /**
+     * Returns the structure that contains the result of BF kNN model-based training
+     * \return Structure that contains the result of BF kNN model-based training
+     */
+    ResultPtr getResult() { return Result::cast(_result); }
+
+    /**
+     * Resets the results of BF kNN model training algorithm
+     */
+    services::Status resetResult() DAAL_C11_OVERRIDE
+    {
+        _result.reset(new ResultType());
+        DAAL_CHECK(_result, services::ErrorNullResult);
+        _res = NULL;
+        return services::Status();
+    }
+
+    /**
+     * Returns a pointer to a newly allocated BF kNN training algorithm
+     * with a copy of the input objects and parameters for this BF kNN training algorithm
+     * in the batch processing mode
+     * \return Pointer to the newly allocated algorithm
+     */
+    services::SharedPtr<Batch<algorithmFPType, method> > clone() const
+    {
+        return services::SharedPtr<Batch<algorithmFPType, method> >(cloneImpl());
+    }
+
+protected:
+    virtual Batch<algorithmFPType, method> * cloneImpl() const DAAL_C11_OVERRIDE
+    {
+        return new Batch<algorithmFPType, method>(*this);
+    }
+
+    services::Status allocateResult() DAAL_C11_OVERRIDE
+    {
+        const ResultPtr res = getResult();
+        DAAL_CHECK(_result, services::ErrorNullResult);
+        services::Status s = res->template allocate<algorithmFPType>((classifier::training::InputIface *)(_in), (ParameterType*)_par, (int)method);
+        _res = _result.get();
+        return s;
+    }
+
+    void initialize()
+    {
+        _ac  = new __DAAL_ALGORITHM_CONTAINER(batch, BatchContainer, algorithmFPType, method)(&_env);
+        _result.reset(new ResultType());
+        _in = &_input;
+        _par = &_param;
+    }
+    InputType _input;
+    ParameterType _param;
```

Comment:
Why `_param` is defined here? There is `_par` defined in the base class. You have to initialize it in constructor with instance of `ParameterType`, line 179 is even more strange to me.